### PR TITLE
amrex::Parser

### PR DESCRIPTION
**Pull request by WeiqunZhang**
Replace WarpXParser with amrex::Parser. Roundoff errors are expected because
of additional optimization in amrex::Parser.

### CODE REVIEW

File: Source/Diagnostics/FlushFormats/FlushFormatPlotfile.cpp [hunk 1]
```diff
@@ -355,7 +354,7 @@ FlushFormatPlotfile::WriteParticles(const std::string& dir,
         UniformFilter const uniform_filter(particle_diags[i].m_do_uniform_filter,
                                            particle_diags[i].m_uniform_stride);
         ParserFilter parser_filter(particle_diags[i].m_do_parser_filter,
-                                   getParser(particle_diags[i].m_particle_filter_parser),
+                                   compileParser<7>(particle_diags[i].m_particle_filter_parser.get()),
```
Comment (ax3l) [inline Source/Diagnostics/FlushFormats/FlushFormatPlotfile.cpp]: Is the `7` here accidentally hard-coded?
Do you want to use a `m_nvars` var as in the other locations?

File: Source/Diagnostics/WarpXOpenPMD.cpp [hunk 2]
```diff
@@ -459,7 +458,7 @@ WarpXOpenPMDPlot::WriteOpenPMDParticles (const amrex::Vector<ParticleDiag>& part
       UniformFilter const uniform_filter(particle_diags[i].m_do_uniform_filter,
                                          particle_diags[i].m_uniform_stride);
       ParserFilter parser_filter(particle_diags[i].m_do_parser_filter,
-                                 getParser(particle_diags[i].m_particle_filter_parser),
+                                 compileParser<7>(particle_diags[i].m_particle_filter_parser.get()),
```
Comment (ax3l) [inline Source/Diagnostics/WarpXOpenPMD.cpp]: Do you want to use a `m_nvars` variable here as well to name the constant?

File: Source/EmbeddedBoundary/WarpXInitEB.cpp [hunk 3]
```diff
@@ -66,7 +66,7 @@ namespace {
         }
 
     private:
-        HostDeviceParser<3> m_parser;
+        amrex::ParserExecutor<3> m_parser;
```
Comment (ax3l) [inline Source/EmbeddedBoundary/WarpXInitEB.cpp]: Probably a good chance to describe the constant:
```suggestion
        amrex::ParserExecutor<3> m_parser; //! function parser with three arguments (x,y,z)
```

**Review by ax3l — COMMENTED** [review]
Thank you for generalizing this! Thanks also for updating the docs.
Awesome that this also resolved the recursive function usage. Also looking forward to the temporary variable support with `r=sqrt(x*x+z*x); ...` expressions :tada: Do you want to add this in one of the tests?

I quickly checked if #1874 still works and it looks like it's covered in CI.

Added minor inline comments, looks already great to me.

File: Source/Diagnostics/FlushFormats/FlushFormatPlotfile.cpp (hunk 1 above)
Comment (WeiqunZhang) [reply Source/Diagnostics/FlushFormats/FlushFormatPlotfile.cpp]: It has always been hard-coded.  I can change it.

**Comment by WeiqunZhang** [comment]
Maybe we can add the temporary variable to tests later.

**Comment by WeiqunZhang** [comment]
As @dpgrote requested, integer is supported in `amerx::IParser`.  The main difference is how division works.  Dave will use it to parse integer parameters in a future PR.

**Comment by WeiqunZhang** [comment]
Not sure why the DPC++ build fails now. :thinking:

**Comment by WeiqunZhang** [comment]
Looks like the issue is `Source/Particles/Collision/BackgroundMCCCollision.cpp` in 437a53f956fd.

**Comment by WeiqunZhang** [comment]
The issue is `MCCProcess` is not trivially copyable because it has a `ManagedVector` member.  I think we should try to fix this in another PR.

**Comment by WeiqunZhang** [comment]
The MCCProcess issue is addressed in #2085.

File: Source/Diagnostics/ReducedDiags/FieldReduction.H [hunk 4]
```diff
@@ -127,7 +127,7 @@ public:
         }
 
         // get parser
-        HostDeviceParser<m_nvars> reduction_function_parser = getParser(m_parser);
+        auto reduction_function_parser = m_parser->compile<m_nvars>();
```
Comment (ax3l) [inline Source/Diagnostics/ReducedDiags/FieldReduction.H:130]: I am wondering if the many specializations in various TUs impact compile-time.
Will take a look when CI ran through just to get an idea.

Comment (ax3l) [reply Source/Diagnostics/ReducedDiags/FieldReduction.H:130]: Haha, the error bar in CI is about 5min wide, thus it's hard to say. Looks still the same, given that fluctuation :D

**Review by ax3l — APPROVED** [review]

**Pull request merged**